pkg/database/user_report.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-server/pkg/base64util"
+	"github.com/jinzhu/gorm"
+)
+
+const (
+	NonceLength = 256
+)
+
+// UserReport is used to de-duplicate phone numbers for user-initiated reporting.
+type UserReport struct {
+	Errorable
+
+	// ID is an auto-increment primary key
+	ID uint
+
+	// PhoneHash is the base64 encoded HMAC of the phone number used to create a user report
+	PhoneHash string `json:"-"` // unique
+	// Nonce is the random data that must be presented when verifying a verification code attached to this user report
+	Nonce string
+
+	// CodeClaimed is set to true when the associated code is claimed. This is needed
+	// since the verification code itself will be cleaned up before this record.
+	CodeClaimed bool
+
+	CreatedAt time.Time
+	UpdatedAt time.Time
+}
+
+// NewUserReport creates a new UserReport by calculating the current HMAC of the
+// provided phone number and encoding the nonce. It does NOT save it to the database.
+func (db *Database) NewUserReport(phone string, nonce []byte) (*UserReport, error) {
+	hmac, err := db.GeneratePhoneNumberHMAC(phone)
+	if err != nil {
+		return nil, err
+	}
+
+	nonceB64 := base64.StdEncoding.EncodeToString(nonce)
```

Comment:
General comment throughout this PR: there's a lot of b64 encoding/decoding using the stdlib. I'd prefer if we used our helper libraries since it more gracefully handles padding/lack-of padding and URL vs std encoding.

Reply:
our helper library is for universal decode - which is always used in this pr.